### ISSUE TITLE
Improperly closed tags break output

### ISSUE THREAD

**Issue by AlenToma** [taoqf/node-html-parser#152]
I am unable to parse the current html from some reason

```html
<div>
<div id="chr-content">
<span>
  lkjasdkjasdkljakldj
</div>
</div>
`
```

## My current Code is

```js
  const validate =()=> {
       html = html.replace(/<!DOCTYPE html>/g, "").replace(/[[class]]/g, "").replace(/[[id]]/g, "")
       var container = parse("<div>" + html + "</div>");
       var content = container.querySelector("#chr-content");
       console.log(content != null ? "found" : "not found")
  }
```

This is on Node.js with the latest version 4.1.4

Here is snack example I created that contain the problem if you would like to test it and see.

[snack](https://snack.expo.dev/@alentoma/html-test)

Right now I am using `react-native-html-parser` together with this library to be able to fix the incorrect html that contains no end tags.

it seems that `node-html-parser` simple ignore and rewrite the html and remove `<div id="chr-content">` from some reason.

----

# Contributor's Note

Although this library was built with the known limitation of requiring proper HTML, we are looking at revising the logic in a way which will not impact performance but will be able to more reasonably handle issues of unmatched open and close tags.

This issue will be left open until that has been addressed

— @nonara

**Comment by taoqf** [comment]
This lib is not suppose to deal with incorrect html. I am so sorry for that. If you could fix this, I am happy to merge you pr.

**Comment by AlenToma** [comment]
Why close the issue then.

I made a simple tidy class the could fix this issue and auto close those open tags.

But there is to many test has to be made to be able to predict all future html.

The class below is able to fix the above example and auto close the span.

```
export default class htmlTydy {
  html: string;
  linesOfCode: string[];
  selfClosedTags: string[] = [
    'area',
    'base',
    'br',
    'col',
    'embed',
    'hr',
    'img',
    'input',
    'link',
    'meta',
    'param',
    'source',
    'track',
    'wbr',
  ];
  n: number;
  stack: { line: number; index: number; tag: string }[] = [];
  constructor(html: string) {
    this.html = html;
    this.linesOfCode = html.split('\n');
    this.n = this.linesOfCode.length;
  }

  last() {
    return this.stack.length <= 0 ? '' : this.stack[this.stack.length - 1].tag;
  }

  lastStack() {
    return this.stack.length <= 0
      ? undefined
      : this.stack[this.stack.length - 1];
  }

  fix() {
    var reg = new RegExp(/[a-zA-Z]/, 'gi');
    var noneChar = new RegExp(/[^a-zA-Z]/, 'gi');
    for (var i = 0; i < this.n; i++) {
      var j = 0;
      var line = this.linesOfCode[i] as string;
      while (j < line.length) {
        var tag = '';
        if (j + 1 < line.length && line[j] == '<' && line[j + 1] == '/') {
          j += 2;
          while (j < line.length && line[j] != '>') {
            tag += line[j];
            j++;
          }
          if (this.last() != tag) {
            tag = '';
            var lst = this.lastStack();
            this.stack.shift();
          } else this.stack.pop();
        } else if (j + 1 < line.length && line[j] == '<' && line[j] == '!')
          continue;
        else if (line[j] == '<') {
          tag = '';
          j++;
          while (j < line.length && line[j] != '>') {
            tag += line[j++];
          }
          if (!this.selfClosedTags.find((x) => x === tag))
            this.stack.push({ line: i, index: j+1, tag: tag });
        }
        j++;
      }
    }

    function addStr(str: string, index: number, stringToAdd: string) {
      return (
        str.substring(0, index) + stringToAdd + str.substring(index, str.length)
      );
    }
 
    this.stack
      .reduce((arr, x) => {
        if (!arr.find((f) => f.line == x.line && f.index == x.index && x.tag === f.tag))
          arr.push(x);
        return arr;
      }, [] as { line: number; index: number; tag: string }[])
      .forEach((x) => {
        try {
          console.log(x)
          x.tag = `</${x.tag.replace(noneChar, '')}>`;
          this.linesOfCode[x.line] = addStr(
            this.linesOfCode[x.line],
            x.index,
            x.tag
          );
        } catch (error) {
          console.log(error);
        }
      });

    return this.linesOfCode.join('\n');
  }
}

```

Here is a snack example that shows its working 
https://snack.expo.dev/@alentoma/html-test

**Comment by AlenToma** [comment]
I have checket the code and noticed something wrong in the code that couse the output to break. 

Check this line 
```js
// Single error  <div> <h3> </div> handle: Just removes <h3>
				oneBefore.removeChild(last);
```
This will remove the child and its content, why dont we just close it ?

**Comment by AlenToma** [comment]
Hi again.
I inspected the code above and did some test.

I do not know why you really remove the last element since you already found the none closed tags.

Anyway here is a possible solution that worked 

I added `parseNoneClosedTags` option

and changed the code to below 

```js
if (last.parentNode && last.parentNode.parentNode) {
			if (last.parentNode === oneBefore && last.tagName === oneBefore.tagName) {
				// Pair error case <h3> <h3> handle : Fixes to <h3> </h3>
				if (options.parseNoneClosedTags !== true) {
					oneBefore.removeChild(last);
					last.childNodes.forEach((child) => {
						oneBefore.parentNode.appendChild(child);
					});
				}
				stack.pop();
			} else {
				// Single error  <div> <h3> </div> handle: Just removes <h3>
				if (options.parseNoneClosedTags !== true) {
					oneBefore.removeChild(last);
					last.childNodes.forEach((child) => {
						oneBefore.appendChild(child);
					});
				}
			}
		}

``` 

And here is the test for this issue which passed.

```js
const { parse } = require('@test/test-target');

describe('issue 152', function () {
	it('shoud parse attributes right', function () {
		const html = `<div>
<div id="chr-content">
<span>
  lkjasdkjasdkljakldj
</div>
</div>`;
		const expected = `<div>
<div id="chr-content">
<span>
  lkjasdkjasdkljakldj

</span></div></div>`;

		const root = parse(html, { parseNoneClosedTags: true });
		root.toString().should.eql(expected);
		// const div = root.firstChild;
		// div.getAttribute('#input').should.eql('');
		// div.getAttribute('(keyup)').should.eql('applyFilter($event)');
		// div.getAttribute('placeholder').should.eql('Ex. IMEI');
		// root.innerHTML.should.eql(html);
	});
});
```

could you please have a look and let me know if this could work, and event better if it did then please check it in and publish it on npm so I could use it.

**Comment by taoqf** [comment]
merged your code in v5.2.2

**Comment by VityaSchel** [comment]
Please tell me if there is a library that can parse malformed html
